src/main/kotlin/app/revanced/patcher/fingerprint/method/impl/MethodFingerprint.kt

```diff
@@ -40,6 +58,136 @@ abstract class MethodFingerprint(
+         */
+        private fun MethodFingerprint.appMethodsWithSameStrings() : List<ClassAndMethod>? {
+            if (strings != null && strings.count() > 0) {
+                // Only check the first String declared
+                return stringMap[strings.first()]
+            }
+            return null
+        }
+
+        /**
+         * @return all app methods that could match this signature.
+         */
+        private fun MethodFingerprint.appMethodsWithSameSignature() : List<ClassAndMethod> {
+            if (accessFlags == null) return allMethods
+
+            var returnTypeValue = returnType
+            if (returnTypeValue == null) {
+                if (AccessFlags.CONSTRUCTOR.isSet(accessFlags)) {
+                    // Constructors always have void return type
+                    returnTypeValue = "V"
+                } else {
+                    return allMethods
+                }
+            }
+
+            val key = buildString {
+                append(accessFlags)
+                append(returnTypeValue.first())
+                if (parameters != null) appendSignatureKeyParameters(parameters)
+            }
+            return signatureMap[key]!!
+        }
+
+        internal fun clearMethodLookupMap() {
+            allMethods.clear()
+            signatureMap.clear()
+            stringMap.clear()
+        }
+
+        /**
+         * resolve faster, by creating culled lists based on method signature and Strings contained.
+         */
+        internal fun createMethodLookupMap(classes: Iterable<ClassDef>) {
+            fun addMethodToMapList(map: MutableMap<String, MutableList<ClassAndMethod>>,
+                                   key: String, keyValue:  ClassAndMethod) {
+                var list = map[key]
+                if (list == null) {
+                    list = LinkedList()
+                    map[key] = list
+                }
+                list += keyValue
+            }
+
+            for (classDef in classes) {
+                for (method in classDef.methods) {
+                    // Key structure is: (access)(returnType)(optional: parameter types)
+                    val accessFlagsReturnKey = method.accessFlags.toString() + method.returnType.first()
+                    val accessFlagsReturnParametersKey = buildString {
+                        append(accessFlagsReturnKey)
+                        appendSignatureKeyParameters(method.parameterTypes)
+                    }
+                    val classAndMethod = ClassAndMethod(classDef, method)
+
+                    // For signatures with no access or return type specified.
+                    allMethods += classAndMethod
+                    // Access and return type.
+                    addMethodToMapList(signatureMap, accessFlagsReturnKey, classAndMethod)
+                    // Access, return, and parameters.
+                    addMethodToMapList(signatureMap, accessFlagsReturnParametersKey, classAndMethod)
```

Comment:
Currently, the `MethodFingerprint` API does not suggest the user design the fingerprint according to this arbitrary key structure because it is not apparent why it exists in this form and shape. The API has to be refactored so it is clear to the user. I am not sure how that should be accomplished, considering this is a highly abstract and arbitrary way of increasing performance.